File: C11/백준/Gold/27172. 수 나누기 게임/수 나누기 게임.c
#include <stdio.h>

#define MAX 1000001

int x[100000];
int count[MAX];
int score[MAX];

int main() {
    int N;
    scanf("%d", &N);

    for (int i = 0; i < N; i++) {
        scanf("%d", &x[i]);
        count[x[i]] = 1;
    }

    for (int i = 0; i < N; i++) {
        int a = x[i];
        for (int j = a * 2; j < MAX; j += a) {
            if (count[j]) {
                score[a]++;
                score[j]--;
            }
        }
    }

    for (int i = 0; i < N; i++) {
        printf("%d ", score[x[i]]);
    }

    return 0;
}

File: Java/백준/Gold/27172. 수 나누기 게임/수 나누기 게임.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int N = Integer.parseInt(br.readLine());
        int[] x = new int[N];
        StringTokenizer st = new StringTokenizer(br.readLine());
        int max = 0;
        for (int i = 0; i < N; i++) {
            x[i] = Integer.parseInt(st.nextToken());
            if (x[i] > max) max = x[i];
        }

        int[] count = new int[1000001];
        for (int i = 0; i < N; i++) {
            count[x[i]] = 1;
        }

        int[] score = new int[1000001];
        for (int i = 0; i < N; i++) {
            int a = x[i];
            for (int j = a * 2; j <= 1000000; j += a) {
                if (count[j] == 1) {
                    score[a]++;
                    score[j]--;
                }
            }
        }

        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < N; i++) {
            sb.append(score[x[i]]).append(" ");
        }
        System.out.println(sb);
    }
}